File: lib/core/src/main/java/io/loglens/core/LogLens.java
package io.loglens.core;

import java.util.ArrayList;
import java.util.List;

/**
 * 파사드. call site 가 보는 전부.
 *
 * <pre>{@code
 * LogLens.init(BuildConfig.DEBUG);                 // 앱이 빌드타입을 주입한다
 * LogLens.addSink(new LogcatSink());
 *
 * LogLens.i(AppDomain.AUTH, "LOGIN_OK", "uid", uid);
 * LogLens.e(AppDomain.NET, "SOCKET_FAIL", e, "host", host);
 * LogLens.d(AppDomain.CHAT, "ROOM_ENTER");
 * LogLens.i(AppDomain.AUTH, "LOGIN_FAIL", "reason", r, "msg", "비밀번호 불일치");
 * }</pre>
 *
 * <h2>왜 varargs 키-값 쌍인가</h2>
 * 채택되지 않는 로깅 표준은 존재하지 않는 것과 같다. {@code Log.d(TAG, "..")} 대비
 * 마찰이 조금이라도 크면 개발자는 그냥 안 쓴다. 빌더 패턴({@code .field().field().log()})은
 * 타입 안전하지만 장황해서 기각했다.
 *
 * <p>대가: 키-값 짝이 컴파일 타임에 보장되지 않는다. 이건 {@code init(isDebug=true)} 일 때
 * 홀수 인자를 시끄럽게 경고하는 것으로 보완한다 — 개발 중에 반드시 걸린다.
 *
 * <h2>릴리스 게이팅</h2>
 * {@code init()} 을 부르기 전 기본값은 <b>디버그 아님</b>이다. 초기화를 잊었을 때
 * V/D 가 새는 것보다 조용히 빠지는 쪽이 안전하다.
 */
public final class LogLens {

    private static volatile boolean debug = false;      // 안전 기본값
    private static volatile Formatter formatter = new Formatter();
    private static final List<Sink> SINKS = new ArrayList<>();

    private LogLens() {}

    // ── 설정 ────────────────────────────────────────────────────────────────

    /**
     * @param isDebug 앱의 디버그 빌드 여부. <b>앱이 주입해야 한다.</b>
     *                라이브러리 모듈의 {@code BuildConfig.DEBUG} 는 자기 모듈의 빌드타입이라
     *                앱이 릴리스여도 true 일 수 있다 (기획서 §5의 함정).
     */
    public static synchronized void init(boolean isDebug) {
        debug = isDebug;
        formatter = new Formatter(new Masker(), Truncator.DEFAULT_MAX_BYTES, diagnostics());
    }

    /** 마스킹 정책까지 갈아끼우는 초기화. */
    public static synchronized void init(boolean isDebug, Masker masker, int maxBytes) {
        debug = isDebug;
        formatter = new Formatter(masker, maxBytes, diagnostics());
    }

    public static synchronized void addSink(Sink sink) {
        if (sink != null) SINKS.add(sink);
    }

    public static synchronized void clearSinks() {
        for (Sink s : SINKS) s.close();
        SINKS.clear();
    }

    public static boolean isDebug() {
        return debug;
    }

    private static Formatter.Diagnostics diagnostics() {
        return debug ? m -> System.err.println("[LogLens] " + m) : Formatter.Diagnostics.SILENT;
    }

    // ── 로깅 API ────────────────────────────────────────────────────────────

    public static void v(LogDomain d, String event, Object... kv) { log(Level.V, d, event, null, kv); }
    public static void d(LogDomain d, String event, Object... kv) { log(Level.D, d, event, null, kv); }
    public static void i(LogDomain d, String event, Object... kv) { log(Level.I, d, event, null, kv); }
    public static void w(LogDomain d, String event, Object... kv) { log(Level.W, d, event, null, kv); }
    public static void e(LogDomain d, String event, Object... kv) { log(Level.E, d, event, null, kv); }

    /** 예외 오버로드. {@code err=} / {@code at=} 필드로 접혀 들어간다. */
    public static void w(LogDomain d, String event, Throwable t, Object... kv) { log(Level.W, d, event, t, kv); }
    public static void e(LogDomain d, String event, Throwable t, Object... kv) { log(Level.E, d, event, t, kv); }

    // ── 내부 ────────────────────────────────────────────────────────────────

    static void log(Level level, LogDomain domain, String event, Throwable t, Object... kv) {
        if (!level.enabled(debug)) return;          // 릴리스에서 V/D 는 문자열 조립조차 안 한다
        if (domain == null) return;
        List<Sink> sinks;
        Formatter f;
        synchronized (LogLens.class) {
            if (SINKS.isEmpty()) return;
            sinks = new ArrayList<>(SINKS);
            f = formatter;
        }
        String body = f.body(event, kv, t);
        String tag = domain.tag();
        for (Sink s : sinks) {
            try {
                s.write(level, tag, body);
            } catch (Throwable ignored) {
                // 로깅이 앱을 죽이면 안 된다.
            }
        }
    }

    /** 테스트/왕복검증용: 싱크를 거치지 않고 본문만 얻는다. */
    public static String render(String event, Throwable t, Object... kv) {
        return formatter.body(event, kv, t);
    }
}

File: lib/core/src/main/kotlin/io/loglens/core/LogLens.kt
package io.loglens.core

/**
 * 호출부가 쓰는 진입점입니다.
 *
 * ### Kotlin
 * ```
 * LogLens.init(BuildConfig.DEBUG)
 * LogLens.addSink(LogcatSink())
 *
 * LogLens.i(AppDomain.AUTH, "LOGIN_OK", "uid" to uid, "msg" to "로그인 성공")
 * LogLens.e(AppDomain.NET, "SOCKET_FAIL", e, "host" to host)
 * LogLens.d(AppDomain.CHAT, "ROOM_ENTER")
 * ```
 *
 * ### Java
 * ```java
 * LogLens.i(AppDomain.AUTH, "LOGIN_OK", "uid", uid, "msg", "로그인 성공");
 * LogLens.e(AppDomain.NET, "SOCKET_FAIL", e, "host", host);
 * LogLens.d(AppDomain.CHAT, "ROOM_ENTER");
 * ```
 *
 * ## 왜 API 가 두 벌인가
 *
 * Kotlin 에서는 `"uid" to uid` 로 넘길 수 있어서 **키-값 짝이 컴파일 시점에 보장**됩니다.
 * 인자 개수를 홀수로 쓰는 실수 자체가 불가능합니다.
 *
 * 그런데 자바에는 `to` 같은 infix 문법이 없어서, 같은 API 를 쓰려면 필드마다
 * `new Pair<>("uid", uid)` 를 써야 합니다. 자바가 대부분인 프로젝트에서는 마찰이 큽니다.
 * 그래서 자바용으로 기존 가변 인자 형태를 그대로 남겨 뒀습니다.
 *
 * Kotlin 전용 오버로드에는 `@JvmSynthetic` 이 붙어 있어서 자바에서는 아예 보이지 않습니다.
 * 두 오버로드가 자바 쪽에서 충돌하는 걸 막기 위해서입니다.
 *
 * ## 릴리스에서 로그가 새지 않게
 *
 * [init] 을 부르기 전 기본값은 **디버그 아님**입니다. 초기화를 깜빡했을 때
 * V/D 가 새는 것보다 조용히 안 찍히는 쪽이 안전합니다.
 */
object LogLens {

    @Volatile
    private var debug = false                       // 안전한 기본값

    @Volatile
    private var formatter = Formatter()

    private val sinks = mutableListOf<Sink>()

    // ── 설정 ────────────────────────────────────────────────────────────────

    /**
     * @param isDebug 앱이 디버그 빌드인지. **앱이 알려줘야 합니다.**
     *   라이브러리 모듈의 `BuildConfig.DEBUG` 는 그 라이브러리의 빌드 타입이라,
     *   앱을 release 로 빌드해도 true 로 남을 수 있습니다.
     */
    @JvmStatic
    @Synchronized
    fun init(isDebug: Boolean) {
        debug = isDebug
        formatter = Formatter(Masker(), Truncator.DEFAULT_MAX_BYTES, diagnostics())
    }

    /** 마스킹 정책과 길이 한도까지 바꾸는 초기화. */
    @JvmStatic
    @Synchronized
    fun init(isDebug: Boolean, masker: Masker, maxBytes: Int) {
        debug = isDebug
        formatter = Formatter(masker, maxBytes, diagnostics())
    }

    @JvmStatic
    @Synchronized
    fun addSink(sink: Sink) {
        sinks.add(sink)
    }

    @JvmStatic
    @Synchronized
    fun clearSinks() {
        sinks.forEach { it.close() }
        sinks.clear()
    }

    @JvmStatic
    fun isDebug(): Boolean = debug

    private fun diagnostics(): Formatter.Diagnostics =
        if (debug) Formatter.Diagnostics { System.err.println("[LogLens] $it") }
        else Formatter.Diagnostics.SILENT

    // ── 필드 없음 (두 언어 공통) ────────────────────────────────────────────
    // 가변 인자 오버로드가 둘이라, 인자가 없을 때 어느 쪽인지 애매해집니다.
    // 인자 없는 오버로드를 따로 두면 이쪽이 가장 구체적이라 항상 이깁니다.

    @JvmStatic fun v(domain: LogDomain, event: String) = log(Level.V, domain, event, null, null)
    @JvmStatic fun d(domain: LogDomain, event: String) = log(Level.D, domain, event, null, null)
    @JvmStatic fun i(domain: LogDomain, event: String) = log(Level.I, domain, event, null, null)
    @JvmStatic fun w(domain: LogDomain, event: String) = log(Level.W, domain, event, null, null)
    @JvmStatic fun e(domain: LogDomain, event: String) = log(Level.E, domain, event, null, null)

    // ── Java 용: 키와 값을 번갈아 넘깁니다 ──────────────────────────────────

    @JvmStatic fun v(domain: LogDomain, event: String, vararg kv: Any?) = log(Level.V, domain, event, null, kv)
    @JvmStatic fun d(domain: LogDomain, event: String, vararg kv: Any?) = log(Level.D, domain, event, null, kv)
    @JvmStatic fun i(domain: LogDomain, event: String, vararg kv: Any?) = log(Level.I, domain, event, null, kv)
    @JvmStatic fun w(domain: LogDomain, event: String, vararg kv: Any?) = log(Level.W, domain, event, null, kv)
    @JvmStatic fun e(domain: LogDomain, event: String, vararg kv: Any?) = log(Level.E, domain, event, null, kv)

    @JvmStatic
    fun w(domain: LogDomain, event: String, t: Throwable, vararg kv: Any?) =
        log(Level.W, domain, event, t, kv)

    @JvmStatic
    fun e(domain: LogDomain, event: String, t: Throwable, vararg kv: Any?) =
        log(Level.E, domain, event, t, kv)

    // ── Kotlin 용: 키-값 짝이 컴파일 시점에 보장됩니다 ──────────────────────
    // @JvmSynthetic 을 붙여 자바에서는 안 보이게 합니다 (위 오버로드와 충돌 방지).

    @JvmSynthetic
    fun v(domain: LogDomain, event: String, vararg fields: Pair<String, Any?>) =
        log(Level.V, domain, event, null, fields.flatten())

    @JvmSynthetic
    fun d(domain: LogDomain, event: String, vararg fields: Pair<String, Any?>) =
        log(Level.D, domain, event, null, fields.flatten())

    @JvmSynthetic
    fun i(domain: LogDomain, event: String, vararg fields: Pair<String, Any?>) =
        log(Level.I, domain, event, null, fields.flatten())

    @JvmSynthetic
    fun w(domain: LogDomain, event: String, vararg fields: Pair<String, Any?>) =
        log(Level.W, domain, event, null, fields.flatten())

    @JvmSynthetic
    fun e(domain: LogDomain, event: String, vararg fields: Pair<String, Any?>) =
        log(Level.E, domain, event, null, fields.flatten())

    @JvmSynthetic
    fun w(domain: LogDomain, event: String, t: Throwable, vararg fields: Pair<String, Any?>) =
        log(Level.W, domain, event, t, fields.flatten())

    @JvmSynthetic
    fun e(domain: LogDomain, event: String, t: Throwable, vararg fields: Pair<String, Any?>) =
        log(Level.E, domain, event, t, fields.flatten())

    // ── 내부 ────────────────────────────────────────────────────────────────

    private fun Array<out Pair<String, Any?>>.flatten(): Array<Any?> {
        val out = arrayOfNulls<Any?>(size * 2)
        forEachIndexed { idx, (k, v) ->
            out[idx * 2] = k
            out[idx * 2 + 1] = v
        }
        return out
    }

    private fun log(
        level: Level,
        domain: LogDomain,
        event: String,
        t: Throwable?,
        kv: Array<out Any?>?,
    ) {
        // 릴리스에서 V/D 는 문자열을 만들지도 않습니다.
        if (!level.enabled(debug)) return

        val snapshot: List<Sink>
        val f: Formatter
        synchronized(this) {
            if (sinks.isEmpty()) return
            snapshot = ArrayList(sinks)
            f = formatter
        }

        val body = f.body(event, kv, t)
        val tag = domain.tag()
        for (s in snapshot) {
            try {
                s.write(level, tag, body)
            } catch (ignored: Throwable) {
                // 로그 때문에 앱이 죽으면 안 됩니다.
            }
        }
    }

    /** 테스트와 대조 검증용: 출력 대상을 거치지 않고 본문만 얻습니다. */
    @JvmStatic
    @JvmOverloads
    fun render(event: String, t: Throwable? = null, vararg kv: Any?): String =
        formatter.body(event, kv, t)
}
